pkg/cache/cluster.go

```diff
@@ -615,6 +630,25 @@ func (c *clusterCache) watchEvents(ctx context.Context, api kube.APIResourceInfo
 				if errors.IsNotFound(err) {
 					c.stopWatching(api.GroupKind, ns)
 				}
+				var connectionUpdated bool
+				if err != nil {
+					if c.connectionStatus != ConnectionStatusFailed {
+						c.log.Info("unable to access cluster", "cluster", c.config.Host, "reason", err.Error())
+						c.connectionStatus = ConnectionStatusFailed
+						connectionUpdated = true
+					}
+				} else if c.connectionStatus != ConnectionStatusSuccessful {
+					c.connectionStatus = ConnectionStatusSuccessful
+					connectionUpdated = true
+				}
```

Comment:
The `clusterCache` lock likely needs to be owned before modifying `c.connectionStatus` within this method. We just need to make sure the lock is released before calling `c.Invalidate()`, as default go locks are not reentrant.

Reply:
Updated :+1:  I haven't used the locks for updating [this block](https://github.com/argoproj/gitops-engine/pull/559/files#diff-9c9e197d543705f08c9b1bc2dc404a55506cfc2935a988e6007d248257aadb1aR851) because I noticed that the locks were acquired in the parent function of sync().